File: vihChatBot/src/main/java/com/vihmessenger/vihchatbot/viewmodel/HomeViewModel.kt
package com.vihmessenger.vihchatbot.viewmodel

import BaseActivity
import android.util.Log
import androidx.lifecycle.MutableLiveData
import com.vihmessenger.vihchatbot.api.services.ApiClient
import com.vihmessenger.vihchatbot.data.model.ChatListModelResponse
import com.vihmessenger.vihchatbot.data.model.EnterPriseDiscoverModel
import com.vihmessenger.vihchatbot.data.model.IndustryResponse
import com.vihmessenger.vihchatbot.data.model.EmailLoginRequest
import com.vihmessenger.vihchatbot.data.model.EmailLoginResponse
import com.vihmessenger.vihchatbot.data.model.SdkFeatureModel
import com.vihmessenger.vihchatbot.data.model.UserProfileRequest
import com.vihmessenger.vihchatbot.data.model.UserProfileResponse
import com.vihmessenger.vihchatbot.data.repository.ChatRepository
import com.vihmessenger.vihchatbot.data.repository.EnterprisesDiscoverRepository
import com.vihmessenger.vihchatbot.data.repository.HomeRepository
import com.vihmessenger.vihchatbot.data.repository.NoConnectionException
import com.vihmessenger.vihchatbot.data.services.ApiService
import com.vihmessenger.vihchatbot.data.services.BaseCloudAPIService
import kotlinx.coroutines.launch

class HomeViewModel(baseActivity: BaseActivity?) : BaseViewModel() {

    private val customerHomeRepository = HomeRepository(
        BaseCloudAPIService.getApiService(ApiService::class.java), baseActivity
    )

    private val homeRepository = HomeRepository(
        ApiClient.apiService, baseActivity
    )

    private val chatRepository = ChatRepository(
        ApiClient.apiService, baseActivity
    )

    private val enterprisesDiscoverRepository =
        EnterprisesDiscoverRepository(ApiClient.apiService, baseActivity)


    internal val sdkFeatureLiveData = MutableLiveData<SdkFeatureModel>()
    val userprofileLiveData = MutableLiveData<UserProfileResponse>()
    val emailLoginLiveData = MutableLiveData<EmailLoginResponse>()
    val emailLoginErrorLiveData = MutableLiveData<String>()

    internal val chatListLiveData = MutableLiveData<ChatListModelResponse>()
    internal val enterprisesDiscoverListLiveData = MutableLiveData<EnterPriseDiscoverModel>()

    internal val industryListLiveData = MutableLiveData<IndustryResponse>()
    private var _selectedIndustries: String = ""

    internal val errorListLiveData = MutableLiveData<String>()

    internal val fragmentTransitionLiveData = MutableLiveData<String>()


    fun getSdkFeatures(showBlockingLoader: Boolean, hashCode: String) {
        scope.launch {
            try {
                sdkFeatureLiveData.postValue(
                    customerHomeRepository.getCustomerHome(
                        showBlockingLoader, hashCode = hashCode
                    ).data
                )
            } catch (e: Throwable) {
                Log.d(TAG, "getSdkFeatures: $e")
                e.printStackTrace()
            }
        }
    }

    fun getUserProfile(showBlockingLoader: Boolean, userProfileRequest: UserProfileRequest) {
        scope.launch {
            try {
                userprofileLiveData.postValue(
                    homeRepository.createUserProfile(
                        showBlockingLoader, userProfileRequest
                    )
                )
            } catch (e: Throwable) {
                e.printStackTrace()
            }
        }
    }

    /** Exchanges a verified Cognito ID token (+ mobile for delivery) for app-session tokens. */
    fun emailLogin(showBlockingLoader: Boolean, request: EmailLoginRequest) {
        scope.launch {
            try {
                emailLoginLiveData.postValue(
                    homeRepository.emailLogin(showBlockingLoader, request)
                )
            } catch (e: Throwable) {
                emailLoginErrorLiveData.postValue(e.message ?: "Email login failed")
                e.printStackTrace()
            }
        }
    }

    fun getChattingListResponse(showBlockingLoader: Boolean, hashcode: String) {
        scope.launch {
            try {
                chatListLiveData.postValue(
                    chatRepository.getChatListResponse(
                        showBlockingLoader, hashcode = hashcode
                    )
                )
            } catch (e: NoConnectionException) {
                // A server error / timeout here must not crash the app — the repo
                // rethrows failures (incl. HTTP 500) as NoConnectionException, and
                // scope uses a plain Job with no handler, so an uncaught throw here
                // reaches the default uncaught handler and kills the process.
                errorListLiveData.postValue(e.message)
            } catch (e: Throwable) {
                errorListLiveData.postValue(e.message)
            }
        }
    }

    fun getEnterpriseDiscoverListResponse(
        showBlockingLoader: Boolean,
        hashCode: String,
        page: Int,
        search: String,
        industries: String
    ) {
        scope.launch {
            try {
                enterprisesDiscoverListLiveData.postValue(
                    enterprisesDiscoverRepository.getEnterpriseDiscoverListResponse(
                        showBlockingLoader, hashCode, page, search, industries
                    )
                )
            } catch (e: NoConnectionException) {
                errorListLiveData.postValue(e.message)
            } catch (e: Throwable) {
                errorListLiveData.postValue(e.message)
            }
        }
    }

    fun getIndustriesListResponse(showBlockingLoader: Boolean) {
        scope.launch {
            try {
                industryListLiveData.postValue(
                    enterprisesDiscoverRepository.getIndustriesResponse(
                        showBlockingLoader
                    )
                )
            } catch (e: Throwable) {
                errorListLiveData.postValue(e.message)
            }
        }
    }

    fun getSelectedIndustries(): String {
        return _selectedIndustries
    }

    fun setSelectedIndustries(industries: String) {
        _selectedIndustries = industries
    }
}

File: Sources/VihChatBotSDK/ViewModels/HomeViewModel.swift
import Foundation

/// Mirrors `viewmodel/HomeViewModel.kt`.
public final class HomeViewModel: BaseViewModel {

    private let homeRepository: HomeRepository
    private let chatRepository: ChatRepository
    private let enterprisesDiscoverRepository: EnterprisesDiscoverRepository

    public let sdkFeatureLiveData = LiveData<SdkFeatureModel>()
    public let userProfileLiveData = LiveData<UserProfileResponse>()
    public let emailLoginLiveData = LiveData<EmailLoginResponse>()
    public let emailLoginErrorLiveData = SingleLiveEvent<String>()
    public let chatListLiveData = LiveData<ChatListModelResponse>()
    public let enterprisesDiscoverListLiveData = LiveData<EnterPriseDiscoverModel>()
    public let industryListLiveData = LiveData<IndustryResponse>()
    public let errorListLiveData = SingleLiveEvent<String>()
    public let fragmentTransitionLiveData = SingleLiveEvent<String>()

    private var _selectedIndustries: String = ""

    public init(loaderHost: LoaderHost?) {
        let api = APIClient.shared.apiService
        self.homeRepository = HomeRepository(apiService: api, loaderHost: loaderHost)
        self.chatRepository = ChatRepository(apiService: api, loaderHost: loaderHost)
        self.enterprisesDiscoverRepository = EnterprisesDiscoverRepository(apiService: api, loaderHost: loaderHost)
        super.init()
    }

    public func getSdkFeatures(showBlockingLoader: Bool, hashCode: String) {
        launch { [weak self] in
            guard let self = self else { return }
            do {
                let response = try await self.homeRepository.getCustomerHome(
                    showBlockingLoader: showBlockingLoader, hashCode: hashCode
                )
                self.sdkFeatureLiveData.postValue(response.data)
                // Persist for theme + welcome lookup (mirrors Android `getVihSettings()`).
                if let encoded = try? JSONEncoder().encode(response.data),
                   let json = String(data: encoded, encoding: .utf8) {
                    VihChatBotSDK.shared.prefs?.vihSettings = json
                }
                DynamicThemeManager.shared.apply(from: response.data)
            } catch {
                CorrelationLogger.warn(message: "getSdkFeatures: \(error.localizedDescription)")
            }
        }
    }

    public func getUserProfile(showBlockingLoader: Bool, request: UserProfileRequest) {
        launch { [weak self] in
            guard let self = self else { return }
            do {
                let response = try await self.homeRepository.createUserProfile(
                    showBlockingLoader: showBlockingLoader, body: request
                )
                self.userProfileLiveData.postValue(response)
            } catch {
                CorrelationLogger.warn(message: "createUserProfile failed", error: error)
            }
        }
    }

    /// Exchanges a verified Cognito ID token (+ mobile for delivery) for app-session tokens.
    public func emailLogin(showBlockingLoader: Bool, request: EmailLoginRequest) {
        launch { [weak self] in
            guard let self = self else { return }
            do {
                let response = try await self.homeRepository.emailLogin(
                    showBlockingLoader: showBlockingLoader, body: request
                )
                self.emailLoginLiveData.postValue(response)
            } catch {
                self.emailLoginErrorLiveData.postValue(error.localizedDescription)
            }
        }
    }

    public func getChattingListResponse(showBlockingLoader: Bool, hashcode: String) {
        launch { [weak self] in
            guard let self = self else { return }
            do {
                let response = try await self.chatRepository.getChatListResponse(
                    showBlockingLoader: showBlockingLoader, hashcode: hashcode
                )
                self.chatListLiveData.postValue(response)
            } catch {
                CorrelationLogger.warn(message: "chatList failed", error: error)
            }
        }
    }

    public func getEnterpriseDiscoverListResponse(
        showBlockingLoader: Bool,
        hashCode: String,
        page: Int,
        search: String,
        industries: String
    ) {
        launch { [weak self] in
            guard let self = self else { return }
            do {
                let response = try await self.enterprisesDiscoverRepository
                    .getEnterpriseDiscoverListResponse(
                        showBlockingLoader: showBlockingLoader,
                        hashCode: hashCode, page: page, search: search, industries: industries
                    )
                self.enterprisesDiscoverListLiveData.postValue(response)
            } catch {
                self.errorListLiveData.postValue(error.localizedDescription)
            }
        }
    }

    public func getIndustriesListResponse(showBlockingLoader: Bool) {
        launch { [weak self] in
            guard let self = self else { return }
            do {
                let response = try await self.enterprisesDiscoverRepository.getIndustriesResponse(
                    showBlockingLoader: showBlockingLoader
                )
                self.industryListLiveData.postValue(response)
            } catch {
                self.errorListLiveData.postValue(error.localizedDescription)
            }
        }
    }

    public func selectedIndustries() -> String { _selectedIndustries }
    public func setSelectedIndustries(_ industries: String) { _selectedIndustries = industries }
}
